fhirpipe/references/reference_binder.py

```diff
@@ -0,0 +1,123 @@
+            sub_fhir_object = self.find_sub_fhir_object(instance, reference_path)
+
+            # If we have a list of references, we want to bind all of them.
+            # Thus, we loop on all the items in sub_fhir_object.
+            if not isinstance(sub_fhir_object, list):
+                sub_fhir_object = [sub_fhir_object]
+
+            for sub in sub_fhir_object:
+                identifier = sub["identifier"]
+                reference_type = sub["type"]
+
+                # Get the map for the needed type
+                map = self.get_collection_map(reference_type)
+                value = identifier["value"]
+                # TODO system should have been automatically filled if needed
+                system = identifier["system"] if "system" in identifier else ""
+
+                # If the identifier is in the map, we can fill the reference
+                if (value, system) in map:
+                    fhir_id = map[(value, system)]
+                    sub["reference"] = f"{reference_type}/{fhir_id}"
+                else:
+                    logging.warning(
+                        f"Could perform reference binding to the resource of type {reference_type} "
+                        f"and identifier {(value, system)}."
+                    )
+
+        self.fhirstore.update(instance["resourceType"], instance["id"], instance)
+
+    def get_collection_map(self, fhir_type):
+        """ Get a dict {(value, system): fhir_id, ...} for the specified resource type.
+        If the type is in identifiers_store, we simply return self.indentifiers_store[fhir_type].
+        Else, we fetch all the resources in the store having an identifier
+        and fill the store before.
+        """
+        if fhir_type not in self.indentifiers_store:
+            # TODO don't do this with the mongo client but with fhirstore or API
+            db_client = get_mongo_client()[fhirpipe.global_config["fhirstore"]["database"]]
+            collection = db_client[fhir_type].find(
+                {"identifier": {"$elemMatch": {"value": {"$exists": True}}}},
+                ["id", "identifier.value", "identifier.system"],
+            )
+
+            identifier_map = {}
+            for doc in collection:
+                for identifier in doc["identifier"]:
+                    value = identifier["value"]
+                    # TODO system should have been automatically filled if needed
+                    system = identifier["system"] if "system" in identifier else ""
+                    identifier_map[(value, system)] = doc["id"]
+
+            self.indentifiers_store[fhir_type] = identifier_map
+
+        return self.indentifiers_store[fhir_type]
+
+    def add_reference(self, resource_mapping, path):
+        """ When a reference is found, this method stores the useful information about it
+        to later perform the binding.
+        """
+        self.map_resource_type[resource_mapping["id"]] = resource_mapping["definitionId"]
```

Comment:
should references use the definitionId (eg: FrenchPatient) rather than definition.type (eg: Patient) ?

Reply:
Oh, I forgot that definitionId was the profile. I have something to fix then.
But I think using FrenchPatient would be better because we wouldn't have to store the (identifier: id) for all Patients but only for FrenchPatient. Do you agree?

Reply:
mhh yeah but you can build a fhir reference with the definitionId. If you can use the definitionId in order to optimize here but use definition.type when building the reference i'm ok with that

Reply:
Do you think I should add definitionId in the meta field of an instance?

Reply:
Oh, yes, we only have definition.type in the mapping.
I'll change that